src/middleware/etaghasher.rs

```diff
@@ -0,0 +1,318 @@
+        use header;
+        use Body;
+
+        if !self.filter.filter(req, &res) {
+            return Ok(middleware::Response::Done(res));
+        }
+
+        let e = if let Body::Binary(b) = res.body() {
+            Some(EntityTag::strong(self.hasher.hash(b.as_ref())))
+        } else {
+            None
+        };
+
+        if let Some(etag) = e {
+            if !none_match(&etag, req) {
+                let mut not_modified =
+                    HttpResponse::NotModified().set(header::ETag(etag)).finish();
+
+                // RFC 7232 requires copying over these headers:
+                copy_header(header::CACHE_CONTROL, &res, &mut not_modified);
+                copy_header(header::CONTENT_LOCATION, &res, &mut not_modified);
+                copy_header(header::DATE, &res, &mut not_modified);
+                copy_header(header::EXPIRES, &res, &mut not_modified);
+                copy_header(header::VARY, &res, &mut not_modified);
+
+                return Ok(middleware::Response::Done(not_modified));
+            }
+            etag.to_string()
+                .parse::<header::HeaderValue>()
+                .map(|v| {
+                    res.headers_mut().insert(header::ETAG, v);
+                })
+                .unwrap_or(());
+        }
+        Ok(middleware::Response::Done(res))
+    }
+}
+
+#[inline]
+fn copy_header(h: ::header::HeaderName, src: &HttpResponse, dst: &mut HttpResponse) {
+    if let Some(val) = src.headers().get(&h) {
+        dst.headers_mut().insert(h, val.clone());
+    }
+}
+
+// Returns true if `req` doesn't have an `If-None-Match` header matching `req`.
+#[inline]
+fn none_match<S>(etag: &EntityTag, req: &HttpRequest<S>) -> bool {
+    use header::IfNoneMatch;
+    use httpmessage::HttpMessage;
+    match req.get_header::<IfNoneMatch>() {
+        Some(IfNoneMatch::Items(ref items)) => {
+            for item in items {
+                if item.weak_eq(etag) {
+                    return false;
+                }
+            }
+            true
+        }
+        Some(IfNoneMatch::Any) => false,
+        None => true,
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use header::ETAG;
+    use http::StatusCode;
+    use httpmessage::HttpMessage;
+    use middleware::Middleware;
+    use test::{TestRequest, TestServer};
+
+    const TEST_ETAG: &'static str = "\"a94a8fe5ccb19ba61c4c0873d391e987982fbbd3\"";
+    struct TestState {
+        _state: u32,
+    }
+    fn test_index<S>(_req: HttpRequest<S>) -> &'static str {
+        "test"
+    }
+
+    fn mwres(r: Result<middleware::Response>) -> HttpResponse {
+        match r {
+            Ok(middleware::Response::Done(hr)) => hr,
+            _ => panic!(),
+        }
+    }
+
+    #[test]
+    fn test_default_create_etag() {
+        let mut eh = EtagHasher::new(DefaultHasher, DefaultFilter);
+        let mut req = TestRequest::default().finish();
+        let res = HttpResponse::Ok().body("test");
```

Comment:
I'd suggest to to move `"test"` literal to constant like `TEST_ETAG`

Reply:
Yes!